### PR TITLE
update conflicting loaded packages  text

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

https://jira.autodesk.com/browse/DYN-4387

* moves cancel button to far right (following standard messagebox pattern)
* adds new overload to `IMessageBox`, `DefaultMessageBox` and `DynamoMessageBox` to allow overriding of button names - the number of names must match the number of buttons, and they are applied in order. IE the order of the names in the `MessageBoxButton` parameter is used to determine which name applies to which button.
* names and text are overridden for this particular message according to UX/Content design.
* modifies `DynamoMessageBox` so that it sets a `MessageBoxResult` when any button is clicked, so we don't need to try converting from a nullable bool (`bool?`) with 3 states, to a MessageBoxResult with 5 states (none,yes,no,cancel,ok)... which was not sufficient to cover all buttons.
* handle the case where a the messageBoxResult could be MessageBoxResult.None (closing the dialog without clicking a button) - this usually results in the same result as cancel, but now we don't proceed as if `OK` was returned...

### TODO:
- [x] add more unit tests for DynamoMessageBox
- [x] audit use of MessageBoxService to check that we are dealing correctly with the return results now that `cancel`, `no` and `closing` the dialog do convert based on the dialogButtons present in the particular dialog, but instead are set directly to their messageboxResult type. 
- [ ] https://jira.autodesk.com/browse/DYN-4404 ... probably let morpheus solve this holistically later.

Screen Shot 2021-12-01 at 5 27 05 PM

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

update text of conflicting loaded packages message, and fix message box button order, and fixed issue with custom dynamo message box buttons not returning correct results.

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@aparajit-pratap - unfortunately, while testing this I discovered that because of our wrapped implementation of a dialogResult in a message box - the `cancel` button result is converted to false, and so does the same exact thing as the `No` or `Uninstall loaded packages` button...

So - we can either get rid of the cancel button, or I can look into how to get the working. I'll take a look at it, but I don't want to break any other dialogs at this point.

**Comment by mjkkirschner** [comment]
FYi @OliverEGreen would appreciate your feedback on my changes to DynamoMessageBox.

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml [hunk 1]
```diff
@@ -110,8 +110,21 @@
                            Grid.Row="2"
                            Margin="0,0,20,20"
                            LastChildFill="False">
+                    
+                    <Button Name="CancelButton"
+                            MinWidth="90"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml:115]: We had an issue somewhere saying German translation of cancel would take a lot of space making cancel button looks like cut. I suspect you may run into same issue here, our team has no solution yet. I think the task is https://jira.autodesk.com/browse/DYN-4404

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 2]
```diff
@@ -23,6 +26,15 @@ public partial class DynamoMessageBox : INotifyPropertyChanged
 
         #region Public Properties
         
+        [Obsolete("Do not use, instead use CustomDialogResult")]
+        public new bool? DialogResult
```
Comment (QilongTang) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:30]: You can just delete, this class is introduced in 2.13.0

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml (hunk 1 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml:115]: I'll take a look, but I've made it so the buttons now grow in width, see the screenshot above.

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:30]: thanks @QilongTang I'm actually trying to wrap the base class (window) public property to alert clients and devs not to use this property, we can't remove the underlying property, but I want to note that it should not be used.

Are you okay with me no longer setting it at all, that will make my test pass.

Comment (QilongTang) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:30]: Sure, this is our own new UI class, we can do whatever we wish is clear

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml (hunk 1 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml:115]: @mjkkirschner Neat, then we may come back here when tackling the other one

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:31]: okay done, DialogResult is no longer set, but I kept the `new` method around to hide the base class version, and added some notes to the class summary about how to use it.

**Comment by mjkkirschner** [comment]
https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/905/

**Comment by mjkkirschner** [comment]
serial tests above passed, still waiting on master-5

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 3]
```diff
@@ -159,7 +175,24 @@ public static MessageBoxResult Show(string messageBoxText, string caption)
                 BodyText = messageBoxText
             };
             dynamoMessageBox.ConfigureButtons(MessageBoxButton.OK);
-            return dynamoMessageBox.ConvertResult(dynamoMessageBox.ShowDialog());
+            dynamoMessageBox.ShowDialog();
+            return dynamoMessageBox.CustomDialogResult;
+        }
+
+        internal static MessageBoxResult Show(string messageBoxText, string caption, MessageBoxButton button, IEnumerable<string> buttonNames,
```
Comment (pinzart90) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:182]: how are the `buttonNames` assigned to the window buttons ?
Is there an order defined somewhere ?
Maybe it should be a list<tuple<button_enum, name>> or dictionary<button_enum, name>

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx [hunk 4]
```diff
@@ -1378,11 +1378,13 @@ Do you wish to continue installing it while keeping the packages that are alread
 If not, {2} needs to uninstall {1} to continue but cannot as they are in use. Try restarting {2} and download {0} again.</value>
   </data>
   <data name="MessageForceInstallOrUninstallUponRestart" xml:space="preserve">
-    <value>Package {0} has one or more dependencies that conflict with the following packages: {1}. Dependency conflicts could cause unintended behavior to occur.
-    
-Do you wish to continue installing it while keeping the packages that are already installed?
+    <value>Package {0} has one or more dependencies that conflict with the following package(s):
+
+{1}
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Properties/Resources.en-US.resx:1383]: Does this formatting handle the case where there is more than one package listed one below the other?

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 3 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:182]: as described in the summary, they order just matches the order defined in the ENUM - so OKCancel , "1" ,"2" -> OK:1 Cancel:2.

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx (hunk 4 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Properties/Resources.en-US.resx:1383]: it does not, I will fix that by adding an option to the `JoinPackageNames` method to join with newlines.

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 3 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:182]: Is that acceptable or do you prefer the more explicit parameter?

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 5]
```diff
@@ -168,56 +201,51 @@ public static MessageBoxResult Show(string messageBoxText, string caption)
         /// Shows or hides buttons on the view as necessary.
         /// </summary>
         /// <param name="messageBoxButton"></param>
-        private void ConfigureButtons(MessageBoxButton messageBoxButton)
+        /// /// <param name="buttonNames">names that will be used to override the standard button names.
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:204]: double code comments chars

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 6]
```diff
@@ -168,56 +201,51 @@ public static MessageBoxResult Show(string messageBoxText, string caption)
...
+                        OkButton.Content = buttonNames.ElementAt(0);
+                    }
                     OkButton.Visibility = Visibility.Visible;
                     break;
                 case MessageBoxButton.OKCancel:
+                    if (buttonNames != null && buttonNames.Count() == 2)
+                    {
+                        OkButton.Content = buttonNames.ElementAt(0);
+                        CancelButton.Content = buttonNames.ElementAt(1);
+                    }
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:222]: What if buttonNames count is 1 here, meaning if there is only one custom name, and the other is the default name?

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 7]
```diff
@@ -168,56 +201,51 @@ public static MessageBoxResult Show(string messageBoxText, string caption)
...
                     OkButton.Visibility = Visibility.Visible;
                     CancelButton.Visibility = Visibility.Visible;
                     break;
                 case MessageBoxButton.YesNoCancel:
+                    if (buttonNames != null && buttonNames.Count() == 3)
+                    {
+                        YesButton.Content = buttonNames.ElementAt(0);
+                        NoButton.Content = buttonNames.ElementAt(1);
+                        CancelButton.Content = buttonNames.ElementAt(2);
+                    }
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:232]: same question here

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 6 above)
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:222]: Should we also validate that the number of custom names is equal to the number of buttons and if not, just use the standard names?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:222]: > Should we also validate that the number of custom names is equal to the number of buttons and if not, just use the standard names?


That is what this does currently, at least thats what I thought I wrote 😉  - if the `buttonNames` count does not equal the expected number of buttons, the content is not modified - (it's set to the original names in the xaml)

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 7 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:232]: checkout the xaml for `DynamoMessageBox` - each button has standard content like
` Content="{x:Static p:Resources.GenericTaskDialogOptionCancel}"`

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 6 above)
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:222]: Ah, sorry, I missed that.

**Review by aparajit-pratap — APPROVED** [review]

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs [hunk 8]
```diff
@@ -23,6 +27,15 @@ public partial class DynamoMessageBox : INotifyPropertyChanged
...
         #region Public Properties
         
+        [Obsolete("Do not use, instead use CustomDialogResult. This property will not be set during ShowDialog()")]
+        public new bool? DialogResult
+        {
+            get => base.DialogResult;
+            set => base.DialogResult = value;
+        }
+
+        internal MessageBoxResult CustomDialogResult { get; set; } = MessageBoxResult.None;
```
Comment (pinzart90) [inline src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:37]: is None equivalent to Cancel?
Do we handle None differently anywhere  ?
If not ...can we just use Cancel instead...might be simpler

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 3 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:182]: must of missed the comment, just goes to  show that it is not idiot proof :)
I think it is acceptable for this API (probably not a lot of traffiic)

**Review by pinzart90 — APPROVED** [review]
1 comment then LGTM

File: src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs (hunk 8 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/UI/Prompts/DynamoMessageBox.xaml.cs:37]: `None` occurs when the close button is clicked, Cancel only occurs when the `Cancel` button is clicked. In this PR I have modified our code to usually check for both of them as the same, but that may not always be the intended behavior.

**Comment by mjkkirschner** [comment]
tests passed here, merging this and will send cherry pick.

**Pull request merged**